### PR TITLE
[Scala] add BucketingModule

### PR DESCRIPTION
**Pull request by Ldpe2G**
@javelinjs @CodingCat pls help to review, thanks.

The train log of the LstmBucketing example:
```bash
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/home/ldpe2g/Mxnet/MXNetScala/mxnet/scala-package/examples/target/classes/lib/slf4j-log4j12-1.7.7.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/ldpe2g/Mxnet/MXNetScala/mxnet/scala-package/assembly/linux-x86_64-gpu/target/mxnet-full_2.11-linux-x86_64-gpu-0.11.1-SNAPSHOT.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
2017-09-03 22:23:00,077 [main] [ml.dmlc.mxnet.examples.module.LstmBucketing] [INFO] - Building vocab ...
2017-09-03 22:23:01,047 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - Summary of dataset ==================
2017-09-03 22:23:01,049 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - bucket of len 10 : 19479 samples
2017-09-03 22:23:01,049 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - bucket of len 20 : 19336 samples
2017-09-03 22:23:01,049 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - bucket of len 30 : 12208 samples
2017-09-03 22:23:01,050 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - bucket of len 40 : 3962 samples
2017-09-03 22:23:01,050 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - bucket of len 50 : 845 samples
2017-09-03 22:23:01,050 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - bucket of len 60 : 160 samples
2017-09-03 22:23:01,087 [main] [MXNetJVM] [INFO] - Try loading mxnet-scala from native path.
2017-09-03 22:23:01,087 [main] [MXNetJVM] [INFO] - Try loading mxnet-scala-linux-x86_64-gpu from native path.
2017-09-03 22:23:01,087 [main] [MXNetJVM] [INFO] - Try loading mxnet-scala-linux-x86_64-cpu from native path.
2017-09-03 22:23:01,087 [main] [MXNetJVM] [WARN] - MXNet Scala native library not found in path. Copying native library from the archive. Consider installing the library somewhere in the path (for Windows: PATH, for Linux: LD_LIBRARY_PATH), or specifying by Java cmd option -Djava.library.path=[lib path].
2017-09-03 22:23:01,094 [main] [ml.dmlc.mxnet.util.NativeLibraryLoader] [INFO] - Loading libmxnet-scala.so from /lib/native/ copying to mxnet-scala
2017-09-03 22:23:01,861 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - Summary of dataset ==================
2017-09-03 22:23:01,862 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - bucket of len 10 : 1531 samples
2017-09-03 22:23:01,862 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - bucket of len 20 : 1518 samples
2017-09-03 22:23:01,862 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - bucket of len 30 : 980 samples
2017-09-03 22:23:01,862 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - bucket of len 40 : 322 samples
2017-09-03 22:23:01,862 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - bucket of len 50 : 65 samples
2017-09-03 22:23:01,862 [main] [ml.dmlc.mxnet.examples.module.BucketIo$BucketSentenceIter] [INFO] - bucket of len 60 : 10 samples
2017-09-03 22:23:01,952 [main] [ml.dmlc.mxnet.examples.module.LstmBucketing] [INFO] - Start training ...
2017-09-03 22:23:23,417 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [50]	Speed: 77.93 samples/sec	Train-perplexity=3411.144531
2017-09-03 22:23:41,439 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [100]	Speed: 88.78 samples/sec	Train-perplexity=1992.701172
2017-09-03 22:24:00,954 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [150]	Speed: 81.99 samples/sec	Train-perplexity=1451.561768
2017-09-03 22:24:19,145 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [200]	Speed: 87.96 samples/sec	Train-perplexity=1152.062866
2017-09-03 22:24:38,165 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [250]	Speed: 84.13 samples/sec	Train-perplexity=971.544922
2017-09-03 22:24:58,624 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [300]	Speed: 78.21 samples/sec	Train-perplexity=851.122986
2017-09-03 22:25:18,033 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [350]	Speed: 82.44 samples/sec	Train-perplexity=760.523315
2017-09-03 22:25:37,030 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [400]	Speed: 84.22 samples/sec	Train-perplexity=690.691223
2017-09-03 22:25:52,196 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [450]	Speed: 105.51 samples/sec	Train-perplexity=629.824768
2017-09-03 22:26:12,404 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [500]	Speed: 79.18 samples/sec	Train-perplexity=587.493652
2017-09-03 22:26:30,122 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [550]	Speed: 90.31 samples/sec	Train-perplexity=549.561340
2017-09-03 22:26:45,363 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [600]	Speed: 104.98 samples/sec	Train-perplexity=515.136047
2017-09-03 22:27:01,973 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [650]	Speed: 96.33 samples/sec	Train-perplexity=486.782349
2017-09-03 22:27:19,316 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [700]	Speed: 92.26 samples/sec	Train-perplexity=462.657806
2017-09-03 22:27:38,118 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [750]	Speed: 85.10 samples/sec	Train-perplexity=443.242981
2017-09-03 22:27:54,342 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [800]	Speed: 98.63 samples/sec	Train-perplexity=424.124054
2017-09-03 22:28:13,723 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [850]	Speed: 82.56 samples/sec	Train-perplexity=409.596039
2017-09-03 22:28:34,807 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [900]	Speed: 75.89 samples/sec	Train-perplexity=397.596893
2017-09-03 22:28:51,873 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [950]	Speed: 93.75 samples/sec	Train-perplexity=384.804413
2017-09-03 22:29:09,942 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1000]	Speed: 88.55 samples/sec	Train-perplexity=373.195343
2017-09-03 22:29:27,238 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1050]	Speed: 92.51 samples/sec	Train-perplexity=362.466766
2017-09-03 22:29:45,571 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1100]	Speed: 87.28 samples/sec	Train-perplexity=352.426636
2017-09-03 22:30:04,561 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1150]	Speed: 84.25 samples/sec	Train-perplexity=343.907318
2017-09-03 22:30:22,923 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1200]	Speed: 87.14 samples/sec	Train-perplexity=335.831635
2017-09-03 22:30:41,965 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1250]	Speed: 84.02 samples/sec	Train-perplexity=328.766724
2017-09-03 22:31:00,414 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1300]	Speed: 86.73 samples/sec	Train-perplexity=321.945465
2017-09-03 22:31:18,668 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1350]	Speed: 87.65 samples/sec	Train-perplexity=315.358246
2017-09-03 22:31:39,950 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1400]	Speed: 75.18 samples/sec	Train-perplexity=310.320557
2017-09-03 22:31:57,051 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1450]	Speed: 93.56 samples/sec	Train-perplexity=304.257355
2017-09-03 22:32:15,983 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1500]	Speed: 84.51 samples/sec	Train-perplexity=299.168793
2017-09-03 22:32:33,629 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1550]	Speed: 90.67 samples/sec	Train-perplexity=294.000275
2017-09-03 22:32:50,973 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1600]	Speed: 92.26 samples/sec	Train-perplexity=288.859192
2017-09-03 22:33:09,909 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1650]	Speed: 84.50 samples/sec	Train-perplexity=284.489929
2017-09-03 22:33:25,946 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[0] Batch [1700]	Speed: 99.77 samples/sec	Train-perplexity=279.440613
2017-09-03 22:33:42,917 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[0] Train-perplexity=275.87402
2017-09-03 22:33:42,918 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[0] Time cost=640295
2017-09-03 22:34:31,864 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[0] Validation-perplexity=135.74944
2017-09-03 22:34:52,242 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [50]	Speed: 79.21 samples/sec	Train-perplexity=149.827713
2017-09-03 22:35:10,406 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [100]	Speed: 88.09 samples/sec	Train-perplexity=141.559311
2017-09-03 22:35:30,007 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [150]	Speed: 81.63 samples/sec	Train-perplexity=140.457932
2017-09-03 22:35:47,977 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [200]	Speed: 89.04 samples/sec	Train-perplexity=136.181244
2017-09-03 22:36:07,003 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [250]	Speed: 84.10 samples/sec	Train-perplexity=136.614304
2017-09-03 22:36:27,666 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [300]	Speed: 77.43 samples/sec	Train-perplexity=138.839371
2017-09-03 22:36:47,304 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [350]	Speed: 81.48 samples/sec	Train-perplexity=138.929733
2017-09-03 22:37:06,063 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [400]	Speed: 85.29 samples/sec	Train-perplexity=138.329239
2017-09-03 22:37:21,321 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [450]	Speed: 104.86 samples/sec	Train-perplexity=133.660614
2017-09-03 22:37:41,310 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [500]	Speed: 80.04 samples/sec	Train-perplexity=134.786667
2017-09-03 22:37:59,160 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [550]	Speed: 89.64 samples/sec	Train-perplexity=133.518814
2017-09-03 22:38:14,416 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [600]	Speed: 104.89 samples/sec	Train-perplexity=130.510941
2017-09-03 22:38:31,438 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [650]	Speed: 94.00 samples/sec	Train-perplexity=128.771240
2017-09-03 22:38:48,702 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [700]	Speed: 92.68 samples/sec	Train-perplexity=127.466934
2017-09-03 22:39:07,420 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [750]	Speed: 85.48 samples/sec	Train-perplexity=127.654953
2017-09-03 22:39:23,879 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [800]	Speed: 97.21 samples/sec	Train-perplexity=126.170967
2017-09-03 22:39:43,467 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [850]	Speed: 81.68 samples/sec	Train-perplexity=126.689476
2017-09-03 22:40:04,779 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [900]	Speed: 75.08 samples/sec	Train-perplexity=127.866859
2017-09-03 22:40:22,307 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [950]	Speed: 91.28 samples/sec	Train-perplexity=127.413704
2017-09-03 22:40:40,799 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1000]	Speed: 86.52 samples/sec	Train-perplexity=126.988014
2017-09-03 22:40:58,185 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1050]	Speed: 92.03 samples/sec	Train-perplexity=126.445763
2017-09-03 22:41:16,609 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1100]	Speed: 86.84 samples/sec	Train-perplexity=125.787125
2017-09-03 22:41:35,430 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1150]	Speed: 85.02 samples/sec	Train-perplexity=125.664368
2017-09-03 22:41:53,563 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1200]	Speed: 88.24 samples/sec	Train-perplexity=125.385033
2017-09-03 22:42:12,418 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1250]	Speed: 84.86 samples/sec	Train-perplexity=125.397980
2017-09-03 22:42:31,169 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1300]	Speed: 85.33 samples/sec	Train-perplexity=125.195595
2017-09-03 22:42:49,496 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1350]	Speed: 87.30 samples/sec	Train-perplexity=124.862335
2017-09-03 22:43:10,928 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1400]	Speed: 74.65 samples/sec	Train-perplexity=125.385269
2017-09-03 22:43:27,795 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1450]	Speed: 94.87 samples/sec	Train-perplexity=124.727852
2017-09-03 22:43:46,464 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1500]	Speed: 85.70 samples/sec	Train-perplexity=124.616035
2017-09-03 22:44:04,330 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1550]	Speed: 89.56 samples/sec	Train-perplexity=124.209824
2017-09-03 22:44:21,333 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1600]	Speed: 94.10 samples/sec	Train-perplexity=123.627563
2017-09-03 22:44:40,098 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1650]	Speed: 85.27 samples/sec	Train-perplexity=123.456902
2017-09-03 22:44:55,869 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[1] Batch [1700]	Speed: 101.46 samples/sec	Train-perplexity=122.525131
2017-09-03 22:45:12,977 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[1] Train-perplexity=122.39102
2017-09-03 22:45:12,977 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[1] Time cost=641113
2017-09-03 22:46:02,314 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[1] Validation-perplexity=111.50342
2017-09-03 22:46:22,804 [main] [ml.dmlc.mxnet.Callback$Speedometer] [INFO] - Epoch[2] Batch [50]	Speed: 78.79 samples/sec	Train-perplexity=124.892357
```

### CODE REVIEW

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala [hunk 1]
```diff
@@ -22,9 +22,32 @@ import java.io.IOException
...
-
+import org.slf4j.Logger
 import scala.collection.mutable.ArrayBuffer
 
+object BaseModule {
+  /**
+   * Check that all input names are in symbol's arguments.
+   */
+  @throws(classOf[IllegalArgumentException])
+  def _checkInputNames(symbol: Symbol, names: IndexedSeq[String],
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala]: is it `private`?

File: scala-package/examples/scripts/rnn/run_lstm_bucketing.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+GPUS="0"
+SAVE_MODEL_PATH=./model/lstm
+
+java -Xmx4G -cp $CLASS_PATH \
+	ml.dmlc.mxnetexamples.rnn.LstmBucketing \
+	--data-train $DATA_TRAIN \
+	--data-val $DATA_VAL \
+	--num-epoch $NUM_EPOCH \
+	--gpus $GPUS \
+	--save-model-path $SAVE_MODEL_PATH
```
Comment (javelinjs) [inline scala-package/examples/scripts/rnn/run_lstm_bucketing.sh]: use 2-space indent

**Comment by javelinjs** [comment]
Besides the examples, can we add some unit test cases?

**Comment by Ldpe2G** [comment]
@javelinjs because the high level rnn api of scala-pkg has not been done yet, so it is a bit hard to add test to BucketingModule, for example:  https://github.com/apache/incubator-mxnet/blob/master/tests/python/train/test_bucketing.py

**Pull request merged**